src/pybind/mgr/orchestrator.py

```diff
@@ -291,6 +297,16 @@ def wait(self, completions):
         """
         raise NotImplementedError()
 
+    @_hide_in_features
+    def get_feature_set(self):
+        """Describes which methods this orchestrator implements"""
+        module = self.__class__
+        features = {a: {'available': getattr(Orchestrator, a, None) != getattr(module, a)}
+                    for a in Orchestrator.__dict__
+                    if not a.startswith('_') and not getattr(getattr(Orchestrator, a), '_hide_in_features', False)
```

Comment:
I guess this is for getting which methods every implementation provides, right?

Instead of mixing introspection (`self.__class__`, `getattr`, ...), which always involves an implicit interface, with  decorators, why not sticking to a single pattern?

For example, rather than a decorator for "hiding features", let's provide a decorator for explicitly exposing features or `@tobeimplemented`. This decorator, instead of adding a hidden extra member to methods, may just use a registry, so no need to iterate over internal class items, but the feature registry.

Reply:
> I guess this is for getting which methods every implementation provides, right?

right.

> Instead of mixing introspection (`self.__class__`, `getattr`, ...), which always involves an implicit interface, with decorators, why not sticking to a single pattern?

This decorator is private to the interface. I don't see a big issue with adding this. Instead, I'd rather concentrate on what @bk201 actually needs. 

> For example, rather than a decorator for "hiding features", let's provide a decorator for explicitly exposing features

Indeed, that would be something to think about.

> or `@tobeimplemented`. This decorator, instead of adding a hidden extra member to methods, may just use a registry, so no need to iterate over internal class items, but the feature registry.

That would IMO be overkill for what is needed right now.

Reply:
The client (e.g Dashboard) can cache the feature map, so the method is only called once when the page is loaded.
I agree with @epuertat, a registry of features is simpler and doesn't introduce any new decorators.

Reply:
We have a [feature registry](http://docs.ceph.com/docs/master/mgr/orchestrator_cli/#current-implementation-status) in the docs. Keeping this up-to-date is hard and is often forgotten. I'd rather have something automatically generated.